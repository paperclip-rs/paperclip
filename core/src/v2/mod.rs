//! Core types and traits associated with the
//! [OpenAPI v2 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).

#[cfg(feature = "actix-base")]
mod actix;
mod extensions;
pub mod models;
#[cfg(feature = "codegen")]
mod resolver;
pub mod schema;

#[cfg(feature = "actix-base")]
pub use self::actix::{
    AcceptedJson, CreatedJson, NoContent, OperationModifier, ResponderWrapper, ResponseWrapper,
};

pub use self::models::{DefaultSchema, ResolvableApi};
pub use self::schema::Schema;
pub use paperclip_macros::*;

#[cfg(feature = "codegen")]
use self::resolver::Resolver;
#[cfg(feature = "codegen")]
use crate::error::ValidationError;

#[cfg(feature = "codegen")]
impl<S: Schema + Default> ResolvableApi<S> {
    /// Consumes this API schema, resolves the references and returns
    /// the resolved schema.
    ///
    /// This walks recursively, collects the referenced schema objects,
    /// substitutes the referenced IDs with the pointer to schema objects
    /// and returns the resolved object or an error if it encountered one.
    pub fn resolve(self) -> Result<ResolvableApi<S>, ValidationError> {
        let mut resolver = Resolver::from((
            self.definitions,
            self.paths,
            self.parameters,
            self.responses,
        ));
        resolver.resolve()?;
        Ok(ResolvableApi {
            swagger: self.swagger,
            info: self.info,
            definitions: resolver.defs,
            paths: resolver.paths,
            base_path: self.base_path,
            host: self.host,
            schemes: self.schemes,
            consumes: self.consumes,
            produces: self.produces,
            coders: self.coders,
            support_crates: self.support_crates,
            parameters: resolver.params,
            responses: resolver.resp,
            spec_format: self.spec_format,
            external_docs: self.external_docs,
            security: self.security,
            security_definitions: self.security_definitions,
            tags: self.tags,
        })
    }
}
