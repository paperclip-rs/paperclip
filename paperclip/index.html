<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Paperclip is a OpenAPI code generator for efficient type-safe compile-time checked HTTP APIs in Rust."><title>paperclip - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="paperclip" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0-nightly (f3db63916 2025-06-17)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../paperclip/index.html">paperclip</a><span class="version">0.9.6</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#modules">Crate Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>paperclip</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/paperclip/lib.rs.html#1-34">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Paperclip is a OpenAPI code generator for efficient type-safe
compile-time checked HTTP APIs in Rust.</p>
<p>See the <a href="https://paperclip-rs.github.io/paperclip">website</a> for detailed
documentation and examples.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="actix/index.html" title="mod paperclip::actix">actix</a></dt><dd>Plugin types, traits and macros for actix-web framework.</dd><dt><a class="mod" href="util/index.html" title="mod paperclip::util">util</a></dt><dt><a class="mod" href="v2/index.html" title="mod paperclip::v2">v2</a></dt><dd>Utilities related to the <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md">OpenAPI v2 specification</a>.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.PaperClipError.html" title="enum paperclip::PaperClipError">Paper<wbr>Clip<wbr>Error</a></dt><dd>Global error which encapsulates all related errors.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="type" href="type.PaperClipResult.html" title="type paperclip::PaperClipResult">Paper<wbr>Clip<wbr>Result</a></dt><dd>Generic result used throughout this library.</dd></dl><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="attr" href="attr.api_v2_schema.html" title="attr paperclip::api_v2_schema">api_<wbr>v2_<wbr>schema</a></dt><dd>Converts your struct to support deserializing from an OpenAPI v2
<a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject">Schema</a>
object (<a href="https://paperclip-rs.github.io/paperclip/paperclip/v2/">example</a>). This adds the necessary fields (in addition to your own fields) and implements the
<code>Schema</code> trait for parsing and codegen.</dd></dl></section></div></main></body></html>