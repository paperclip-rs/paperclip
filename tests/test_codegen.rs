use paperclip::v2::{
    self,
    codegen::{CrateMeta, DefaultEmitter, EmitMode, Emitter, EmitterState},
    models::{DefaultSchema, ResolvableApi},
};

use once_cell::sync::Lazy;
use std::{fs::File, path::PathBuf};

static ROOT: Lazy<PathBuf> = Lazy::new(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")));
static PET_SCHEMA: Lazy<ResolvableApi<DefaultSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/pet-v2.yaml")).expect("file?");
    let raw: ResolvableApi<DefaultSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static K8S_SCHEMA: Lazy<ResolvableApi<K8sSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/k8s-v1.16.0-alpha.0-openapi-v2.json")).expect("file?");
    let raw: ResolvableApi<K8sSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static CODEGEN_PET_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_LIB_NO_ROOT: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet/no_root");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    meta.no_root = true;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_pet/cli");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});

static CODEGEN_K8S_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests");
    state.working_dir.push("test_k8s");
    state.mod_prefix = "crate::codegen::";

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});
static CODEGEN_K8S_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_k8s/cli");
    let mut meta = CrateMeta::default();
    assert_eq!(meta.mode, EmitMode::Module);
    meta.name = Some("test-k8s-cli".into());
    meta.version = Some("0.0.0".into());
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
enum PatchStrategy {
    Merge,
    RetainKeys,
    #[serde(rename = "merge,retainKeys")]
    MergeAndRetain,
}

#[paperclip::api_v2_schema]
#[allow(dead_code)]
#[derive(Debug, serde::Deserialize)]
struct K8sSchema {
    #[serde(rename = "x-kubernetes-patch-strategy")]
    patch_strategy: Option<PatchStrategy>,
}

static CODEGEN: Lazy<()> = Lazy::new(|| {
    env_logger::builder()
        .filter(Some("paperclip"), log::LevelFilter::Info)
        .init();
    Lazy::force(&CODEGEN_PET_LIB);
    Lazy::force(&CODEGEN_PET_LIB_NO_ROOT);
    Lazy::force(&CODEGEN_PET_CLI);
    Lazy::force(&CODEGEN_K8S_LIB);
    Lazy::force(&CODEGEN_K8S_CLI);
});

fn assert_file(path: &str) {
    Lazy::force(&CODEGEN);

    let data = std::fs::read_to_string(ROOT.join(path))
        .unwrap_or_else(|err| panic!("missing file {}: {}", path, err));
    insta::assert_snapshot!(path, data);
}

#[cfg(test)]
mod tests_pet {
    use super::*;
    #[test]
    fn test_lib_root() {
        assert_file("tests/test_pet/lib.rs");
    }

    #[test]
    fn test_no_root_manifest() {
        assert_file("tests/test_pet/no_root/Cargo.toml");
    }

    #[test]
    fn test_manifest() {
        assert_file("tests/test_pet/Cargo.toml");
    }

    #[test]
    fn test_header_parameters() {
        assert_file("tests/test_pet/pet.rs");
    }

    #[test]
    fn test_cli_params() {
        assert_file("tests/test_pet/cli/app.yaml");
    }

    #[test]
    fn test_recusrive_object() {
        assert_file("tests/test_pet/recursive_container.rs");
        assert_file("tests/test_pet/recursive_object.rs");
    }

    #[test]
    fn test_anonymous_object_definition_in_schema() {
        // An object "can" define objects in its schema without referencing
        // them from known definitions. In that case, we autogenerate stuff.
        assert_file("tests/test_pet/order.rs");
    }
    #[test]
    fn test_anonymous_object_definition_in_body() {
        assert_file("tests/test_pet/post_shipments_body.rs");
    }

    #[test]
    fn test_anonumous_object_definition_in_response() {
        assert_file("tests/test_pet/get_shipments_id_response.rs");
    }

    #[test]
    fn test_simple_array_parameter_in_path() {
        assert_file("tests/test_pet/status.rs");
    }

    #[test]
    fn test_misc() {
        assert_file("tests/test_pet/miscellaneous.rs");
    }

    #[test]
    fn test_multipart_with_file() {
        assert_file("tests/test_pet/cli/status.rs");
    }

    #[test]
    fn test_simple_any_enum() {
        assert_file("tests/test_pet/test_enum.rs");
    }
}

#[cfg(test)]
mod tests_k8s {
    use super::*;
    use paperclip::v2::models::{HttpMethod, Version};

    #[test]
    fn test_definition_ref_cycles() {
        assert_eq!(K8S_SCHEMA.swagger, Version::V2);
        assert_eq!(K8S_SCHEMA.definitions.len(), 663);

        let json_props_def = &K8S_SCHEMA.definitions
            ["io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps"];
        let desc = json_props_def.read().description.clone();
        let all_of = json_props_def.read().properties["allOf"].clone();
        let items = all_of.read().items.as_ref().unwrap().clone();
        assert_eq!(items.read().description, desc); // both point to same `JSONSchemaProps`
    }

    #[test]
    fn test_resolved_schema() {
        let resp = &K8S_SCHEMA.paths["/api/"].methods[&HttpMethod::Get].responses["200"].read();
        let schema = resp.schema.as_ref().expect("bleh?").read();
        assert!(schema.reference.is_none()); // this was a reference
        assert_eq!(
            &K8S_SCHEMA.definitions["io.k8s.apimachinery.pkg.apis.meta.v1.APIVersions"]
                .read()
                .description
                .as_ref()
                .unwrap()
                .as_str(),
            schema.description.as_ref().unwrap()
        );
    }

    #[test]
    fn test_child_module_declarations() {
        assert_file("tests/test_k8s/io/k8s/api/mod.rs");
    }

    #[test]
    fn test_transparency_with_params() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/custom_resource_definition.rs");
    }
    #[test]
    fn test_struct_for_complex_object() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/json_schema_props.rs");
    }

    #[test]
    fn test_root_mod() {
        assert_file("tests/test_k8s/mod.rs");
    }

    #[test]
    fn test_generics_mod() {
        assert_file("tests/test_k8s/generics.rs");
    }

    #[test]
    fn test_same_object_creates_multiple_builders() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/config_map.rs")
    }

    #[test]
    fn test_same_object_with_multiple_builders_has_basic_builder() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/pod.rs")
    }

    #[test]
    fn test_simple_object_builder_with_required_fields() {
        assert_file("tests/test_k8s/io/k8s/api/rbac/v1/policy_rule.rs")
    }

    #[test]
    fn test_builder_with_field_parameter_collision_and_method_collision() {
        // grace_period_seconds, orphan_dependents and propagation_policy
        // exist in both the object and as a query parameter. If one is set,
        // we should also set the other.
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/delete_options.rs");
    }

    #[test]
    fn test_unit_builder_with_no_modifier() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/api_group_list.rs");
    }

    #[test]
    fn test_builder_field_with_iterators() {
        assert_file(
            "tests/test_k8s/io/k8s/api/certificates/v1beta1/certificate_signing_request_spec.rs",
        );
    }

    #[test]
    fn test_any_in_operation_bound_to_unrelated_struct() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/patch.rs");
    }

    #[test]
    fn test_cli_manifest() {
        assert_file("tests/test_k8s/cli/Cargo.toml");
    }

    #[test]
    fn test_cli_main() {
        assert_file("tests/test_k8s/cli/main.rs");
    }

    #[test]
    fn test_clap_yaml() {
        assert_file("tests/test_k8s/cli/app.yaml");
    }
}
