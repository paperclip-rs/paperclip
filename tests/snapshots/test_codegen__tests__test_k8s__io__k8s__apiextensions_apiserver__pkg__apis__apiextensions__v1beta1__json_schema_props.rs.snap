---
source: tests/test_codegen.rs
expression: data
---

/// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct JsonSchemaProps<Any> {
    #[serde(rename = "$ref")]
    pub ref_: Option<String>,
    #[serde(rename = "$schema")]
    pub schema: Option<String>,
    #[serde(rename = "additionalItems")]
    pub additional_items: Option<Any>,
    #[serde(rename = "additionalProperties")]
    pub additional_properties: Option<Any>,
    #[serde(rename = "allOf")]
    pub all_of: Option<Vec<crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    #[serde(rename = "anyOf")]
    pub any_of: Option<Vec<crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub default: Option<Any>,
    pub definitions: Option<paperclip::v2::PropertiesMap<String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub dependencies: Option<paperclip::v2::PropertiesMap<String, Any>>,
    pub description: Option<String>,
    #[serde(rename = "enum")]
    pub enum_: Option<Vec<Any>>,
    pub example: Option<Any>,
    #[serde(rename = "exclusiveMaximum")]
    pub exclusive_maximum: Option<bool>,
    #[serde(rename = "exclusiveMinimum")]
    pub exclusive_minimum: Option<bool>,
    #[serde(rename = "externalDocs")]
    pub external_docs: Option<crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::external_documentation::ExternalDocumentation>,
    pub format: Option<String>,
    pub id: Option<String>,
    pub items: Option<Any>,
    #[serde(rename = "maxItems")]
    pub max_items: Option<i64>,
    #[serde(rename = "maxLength")]
    pub max_length: Option<i64>,
    #[serde(rename = "maxProperties")]
    pub max_properties: Option<i64>,
    pub maximum: Option<f64>,
    #[serde(rename = "minItems")]
    pub min_items: Option<i64>,
    #[serde(rename = "minLength")]
    pub min_length: Option<i64>,
    #[serde(rename = "minProperties")]
    pub min_properties: Option<i64>,
    pub minimum: Option<f64>,
    #[serde(rename = "multipleOf")]
    pub multiple_of: Option<f64>,
    pub not: Option<Box<crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub nullable: Option<bool>,
    #[serde(rename = "oneOf")]
    pub one_of: Option<Vec<crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub pattern: Option<String>,
    #[serde(rename = "patternProperties")]
    pub pattern_properties: Option<paperclip::v2::PropertiesMap<String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub properties: Option<paperclip::v2::PropertiesMap<String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>>,
    pub required: Option<Vec<String>>,
    pub title: Option<String>,
    #[serde(rename = "type")]
    pub type_: Option<String>,
    #[serde(rename = "uniqueItems")]
    pub unique_items: Option<bool>,
    /// x-kubernetes-embedded-resource defines that the value is an embedded Kubernetes runtime.Object, with TypeMeta and ObjectMeta. The type must be object. It is allowed to further restrict the embedded object. kind, apiVersion and metadata are validated automatically. x-kubernetes-preserve-unknown-fields is allowed to be true, but does not have to be if the object is fully specified (up to kind, apiVersion, metadata).
    #[serde(rename = "x-kubernetes-embedded-resource")]
    pub x_kubernetes_embedded_resource: Option<bool>,
    /// x-kubernetes-int-or-string specifies that this value is either an integer or a string. If this is true, an empty type is allowed and type as child of anyOf is permitted if following one of the following patterns:
    ///
    /// 1) anyOf:
    ///    - type: integer
    ///    - type: string
    /// 2) allOf:
    ///    - anyOf:
    ///      - type: integer
    ///      - type: string
    ///    - ... zero or more
    #[serde(rename = "x-kubernetes-int-or-string")]
    pub x_kubernetes_int_or_string: Option<bool>,
    /// x-kubernetes-preserve-unknown-fields stops the API server decoding step from pruning fields which are not specified in the validation schema. This affects fields recursively, but switches back to normal pruning behaviour if nested properties or additionalProperties are specified in the schema. This can either be true or undefined. False is forbidden.
    #[serde(rename = "x-kubernetes-preserve-unknown-fields")]
    pub x_kubernetes_preserve_unknown_fields: Option<bool>,
}

impl<Any: Default> JsonSchemaProps<Any> {
    /// Create a builder for this object.
    #[inline]
    pub fn builder() -> JsonSchemaPropsBuilder<Any> {
        JsonSchemaPropsBuilder {
            body: Default::default(),
        }
    }
}

impl<Any> Into<JsonSchemaProps<Any>> for JsonSchemaPropsBuilder<Any> {
    fn into(self) -> JsonSchemaProps<Any> {
        self.body
    }
}

/// Builder for [`JsonSchemaProps`](./struct.JsonSchemaProps.html) object.
#[derive(Debug, Clone)]
pub struct JsonSchemaPropsBuilder<Any> {
    body: self::JsonSchemaProps<Any>,
}

impl<Any> JsonSchemaPropsBuilder<Any> {
    #[inline]
    pub fn ref_(mut self, value: impl Into<String>) -> Self {
        self.body.ref_ = Some(value.into());
        self
    }

    #[inline]
    pub fn schema(mut self, value: impl Into<String>) -> Self {
        self.body.schema = Some(value.into());
        self
    }

    #[inline]
    pub fn additional_items(mut self, value: impl Into<Any>) -> Self {
        self.body.additional_items = Some(value.into());
        self
    }

    #[inline]
    pub fn additional_properties(mut self, value: impl Into<Any>) -> Self {
        self.body.additional_properties = Some(value.into());
        self
    }

    #[inline]
    pub fn all_of(mut self, value: impl Iterator<Item = crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>) -> Self {
        self.body.all_of = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn any_of(mut self, value: impl Iterator<Item = crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>) -> Self {
        self.body.any_of = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn default(mut self, value: impl Into<Any>) -> Self {
        self.body.default = Some(value.into());
        self
    }

    #[inline]
    pub fn definitions(mut self, value: impl Iterator<Item = (String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>)>) -> Self {
        self.body.definitions = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }

    #[inline]
    pub fn dependencies(mut self, value: impl Iterator<Item = (String, impl Into<Any>)>) -> Self {
        self.body.dependencies = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }

    #[inline]
    pub fn description(mut self, value: impl Into<String>) -> Self {
        self.body.description = Some(value.into());
        self
    }

    #[inline]
    pub fn enum_(mut self, value: impl Iterator<Item = impl Into<Any>>) -> Self {
        self.body.enum_ = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn example(mut self, value: impl Into<Any>) -> Self {
        self.body.example = Some(value.into());
        self
    }

    #[inline]
    pub fn exclusive_maximum(mut self, value: impl Into<bool>) -> Self {
        self.body.exclusive_maximum = Some(value.into());
        self
    }

    #[inline]
    pub fn exclusive_minimum(mut self, value: impl Into<bool>) -> Self {
        self.body.exclusive_minimum = Some(value.into());
        self
    }

    #[inline]
    pub fn external_docs(mut self, value: crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::external_documentation::ExternalDocumentation) -> Self {
        self.body.external_docs = Some(value.into());
        self
    }

    #[inline]
    pub fn format(mut self, value: impl Into<String>) -> Self {
        self.body.format = Some(value.into());
        self
    }

    #[inline]
    pub fn id(mut self, value: impl Into<String>) -> Self {
        self.body.id = Some(value.into());
        self
    }

    #[inline]
    pub fn items(mut self, value: impl Into<Any>) -> Self {
        self.body.items = Some(value.into());
        self
    }

    #[inline]
    pub fn max_items(mut self, value: impl Into<i64>) -> Self {
        self.body.max_items = Some(value.into());
        self
    }

    #[inline]
    pub fn max_length(mut self, value: impl Into<i64>) -> Self {
        self.body.max_length = Some(value.into());
        self
    }

    #[inline]
    pub fn max_properties(mut self, value: impl Into<i64>) -> Self {
        self.body.max_properties = Some(value.into());
        self
    }

    #[inline]
    pub fn maximum(mut self, value: impl Into<f64>) -> Self {
        self.body.maximum = Some(value.into());
        self
    }

    #[inline]
    pub fn min_items(mut self, value: impl Into<i64>) -> Self {
        self.body.min_items = Some(value.into());
        self
    }

    #[inline]
    pub fn min_length(mut self, value: impl Into<i64>) -> Self {
        self.body.min_length = Some(value.into());
        self
    }

    #[inline]
    pub fn min_properties(mut self, value: impl Into<i64>) -> Self {
        self.body.min_properties = Some(value.into());
        self
    }

    #[inline]
    pub fn minimum(mut self, value: impl Into<f64>) -> Self {
        self.body.minimum = Some(value.into());
        self
    }

    #[inline]
    pub fn multiple_of(mut self, value: impl Into<f64>) -> Self {
        self.body.multiple_of = Some(value.into());
        self
    }

    #[inline]
    pub fn not(mut self, value: crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>) -> Self {
        self.body.not = Some(value.into());
        self
    }

    #[inline]
    pub fn nullable(mut self, value: impl Into<bool>) -> Self {
        self.body.nullable = Some(value.into());
        self
    }

    #[inline]
    pub fn one_of(mut self, value: impl Iterator<Item = crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>>) -> Self {
        self.body.one_of = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn pattern(mut self, value: impl Into<String>) -> Self {
        self.body.pattern = Some(value.into());
        self
    }

    #[inline]
    pub fn pattern_properties(mut self, value: impl Iterator<Item = (String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>)>) -> Self {
        self.body.pattern_properties = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }

    #[inline]
    pub fn properties(mut self, value: impl Iterator<Item = (String, crate::codegen::io::k8s::apiextensions_apiserver::pkg::apis::apiextensions::v1beta1::json_schema_props::JsonSchemaProps<Any>)>) -> Self {
        self.body.properties = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }

    #[inline]
    pub fn required(mut self, value: impl Iterator<Item = impl Into<String>>) -> Self {
        self.body.required = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn title(mut self, value: impl Into<String>) -> Self {
        self.body.title = Some(value.into());
        self
    }

    #[inline]
    pub fn type_(mut self, value: impl Into<String>) -> Self {
        self.body.type_ = Some(value.into());
        self
    }

    #[inline]
    pub fn unique_items(mut self, value: impl Into<bool>) -> Self {
        self.body.unique_items = Some(value.into());
        self
    }

    /// x-kubernetes-embedded-resource defines that the value is an embedded Kubernetes runtime.Object, with TypeMeta and ObjectMeta. The type must be object. It is allowed to further restrict the embedded object. kind, apiVersion and metadata are validated automatically. x-kubernetes-preserve-unknown-fields is allowed to be true, but does not have to be if the object is fully specified (up to kind, apiVersion, metadata).
    #[inline]
    pub fn x_kubernetes_embedded_resource(mut self, value: impl Into<bool>) -> Self {
        self.body.x_kubernetes_embedded_resource = Some(value.into());
        self
    }

    /// x-kubernetes-int-or-string specifies that this value is either an integer or a string. If this is true, an empty type is allowed and type as child of anyOf is permitted if following one of the following patterns:
    ///
    /// 1) anyOf:
    ///    - type: integer
    ///    - type: string
    /// 2) allOf:
    ///    - anyOf:
    ///      - type: integer
    ///      - type: string
    ///    - ... zero or more
    #[inline]
    pub fn x_kubernetes_int_or_string(mut self, value: impl Into<bool>) -> Self {
        self.body.x_kubernetes_int_or_string = Some(value.into());
        self
    }

    /// x-kubernetes-preserve-unknown-fields stops the API server decoding step from pruning fields which are not specified in the validation schema. This affects fields recursively, but switches back to normal pruning behaviour if nested properties or additionalProperties are specified in the schema. This can either be true or undefined. False is forbidden.
    #[inline]
    pub fn x_kubernetes_preserve_unknown_fields(mut self, value: impl Into<bool>) -> Self {
        self.body.x_kubernetes_preserve_unknown_fields = Some(value.into());
        self
    }
}

